cdc/sink/producer/kafka/kafka.go

```diff
@@ -395,80 +409,121 @@ func (k *kafkaSaramaProducer) run(ctx context.Context) error {
-func kafkaTopicPreProcess(topic, address string, config *Config, cfg *sarama.Config) (int32, error) {
-	admin, err := sarama.NewClusterAdmin(strings.Split(address, ","), cfg)
+func topicPreProcess(config *Config, saramaConfig *sarama.Config) error {
+	failpoint.Inject("workaround4Test", func() {
+		failpoint.Return(nil)
+	})
+	admin, err := sarama.NewClusterAdmin(config.BrokerEndpoints, saramaConfig)
 	if err != nil {
-		return 0, cerror.WrapError(cerror.ErrKafkaNewSaramaProducer, err)
+		return err
 	}
 	defer func() {
-		err := admin.Close()
-		if err != nil {
-			log.Warn("close admin client failed", zap.Error(err))
+		if err := admin.Close(); err != nil {
+			log.Warn("close kafka cluster admin failed", zap.Error(err))
 		}
 	}()
+
 	topics, err := admin.ListTopics()
 	if err != nil {
-		return 0, cerror.WrapError(cerror.ErrKafkaNewSaramaProducer, err)
-	}
-	partitionNum := config.PartitionNum
-	topicDetail, exist := topics[topic]
-	if exist {
-		log.Info("get partition number of topic", zap.String("topic", topic), zap.Int32("partition_num", topicDetail.NumPartitions))
-		if partitionNum == 0 {
-			partitionNum = topicDetail.NumPartitions
-		} else if partitionNum < topicDetail.NumPartitions {
-			log.Warn("partition number assigned in sink-uri is less than that of topic", zap.Int32("topic partition num", topicDetail.NumPartitions))
-		} else if partitionNum > topicDetail.NumPartitions {
-			return 0, cerror.ErrKafkaInvalidPartitionNum.GenWithStack(
-				"partition number(%d) assigned in sink-uri is more than that of topic(%d)", partitionNum, topicDetail.NumPartitions)
+		return err
+	}
+
+	var realPartitionCount int32 = 0
+	info, ok := topics[config.TopicName]
+	if ok {
+		realPartitionCount = info.NumPartitions
+	}
+
+	// if user specify cdc to create the topic, we must make sure
+	// that `partition-num` and `replication-factor` is given
+	if config.AutoCreate {
+		// if the specified topic name already exist, we should make sure that
+		// topic's `max.message.bytes` is not less than given `max-message-size`
+		if ok {
+			if a, ok := info.ConfigEntries["max.message.bytes"]; ok {
+				topicMaxMessageBytes, err := strconv.Atoi(*a)
+				if err != nil {
+					return err
+				}
+				if topicMaxMessageBytes < config.MaxMessageBytes {
+					return cerror.ErrKafkaInvalidConfig.GenWithStack(
+						"topic already exist, and max.message.size(%d) less than max-message-size(%d)",
+						topicMaxMessageBytes, config.MaxMessageBytes)
+				}
+			}
+
+			config.PartitionNum = realPartitionCount
+			log.Warn("topic already exist", zap.String("topic", config.TopicName))
+			return nil
 		}
-	} else {
-		if partitionNum == 0 {
-			partitionNum = defaultPartitionNum
-			log.Warn("topic not found and partition number is not specified, using default partition number", zap.String("topic", topic), zap.Int32("partition_num", partitionNum))
+		// this could happen if user does not specify the `partition-num` in the sink uri.
+		if config.PartitionNum == 0 {
```

Comment:
Is it acceptable if the user specifies `partition-num=0`?

Reply:
this could happen if the user does not specify the `partition-num`, which would be 0.

Reply:
Is it checked that the user does not specified `partition-num=0`?

Reply:
we have check `partition-num` in https://github.com/pingcap/ticdc/blob/master/cdc/sink/producer/kafka/kafka.go#L74

But if they does not pass `partition-num` in the sink-uri, it would be 0.